mailbox/event/event-memory/src/main/java/org/apache/james/mailbox/events/MemoryEventDeadLetters.java

```diff
@@ -0,0 +1,92 @@
+/****************************************************************
+ * Licensed to the Apache Software Foundation (ASF) under one   *
+ * or more contributor license agreements.  See the NOTICE file *
+ * distributed with this work for additional information        *
+ * regarding copyright ownership.  The ASF licenses this file   *
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.mailbox.events;
+
+import java.util.Optional;
+
+import org.apache.james.mailbox.Event;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.HashMultimap;
+import com.google.common.collect.Multimap;
+import com.google.common.collect.Multimaps;
+
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+public class MemoryEventDeadLetters implements EventDeadLetters {
+
+    private static final String REGISTERED_GROUP_CANNOT_BE_NULL = "registeredGroup cannot be null";
+    private static final String FAIL_DELIVERED_EVENT_CANNOT_BE_NULL = "failDeliveredEvent cannot be null";
+    private static final String FAIL_DELIVERED_ID_EVENT_CANNOT_BE_NULL = "failDeliveredEventId cannot be null";
+
+    private final Multimap<Group, Event> deadLetters;
+
+    MemoryEventDeadLetters() {
+        this.deadLetters = Multimaps.synchronizedSetMultimap(HashMultimap.create());
+    }
+
+    @Override
+    public Mono<Void> store(Group registeredGroup, Event failDeliveredEvent) {
+        Preconditions.checkArgument(registeredGroup != null, REGISTERED_GROUP_CANNOT_BE_NULL);
+        Preconditions.checkArgument(failDeliveredEvent != null, FAIL_DELIVERED_EVENT_CANNOT_BE_NULL);
+
+        deadLetters.put(registeredGroup, failDeliveredEvent);
+        return Mono.empty();
```

Comment:
It kind of breaks the Reactor paradigm, when a `Mono<>`/`Flux<>` value is returned by a method, we expect that no action is performed, it improves composability.

Here (and in the other methods of the class), you perform the operation first and then you produce an acknowledge value, it will lead to major difference between InMemory and real implementation.